Engine/Helper.cs

```diff
@@ -1348,50 +1348,47 @@ internal List<RuleSuppression> GetSuppressionsConfiguration(ConfigurationDefinit
             }
 
             List<RuleSuppression> ruleSuppressions = ruleSuppressionsDict[ruleName];
-            var offsetArr = GetOffsetArray(diagnostics);
-            int recordIndex = 0;
-            int startRecord = 0;
-            bool[] suppressed = new bool[diagnostics.Count];
-            foreach (RuleSuppression ruleSuppression in ruleSuppressions)
-            {
-                int suppressionCount = 0;
-                while (startRecord < diagnostics.Count
-                    // && diagnostics[startRecord].Extent.StartOffset < ruleSuppression.StartOffset)
-                    // && diagnostics[startRecord].Extent.StartLineNumber < ruleSuppression.st)
-                    && offsetArr[startRecord] != null && offsetArr[startRecord].Item1 < ruleSuppression.StartOffset)
-                {
-                    startRecord += 1;
-                }
+            bool[] applied = new bool[ruleSuppressions.Count];
```

Comment:
From our offline conversation:

> In the past mechanism,if one diagnostic's offset is null and the id matches a rulesuppression id,it will defined as a suppressed record.

Ah, sounds like I've misinterpreted the existing logic. It sounds like we just need to change that part to:

```csharp
if (offsetPair is null)
{
    suppressedRecords.Add(...);
    continue;
}
```

Reply:
when an offsetpair is null, its id also needs to match a rulesuppression id and then will define as a suppressed record. I changed the code to below:  

            foreach (DiagnosticRecord diagnostic in diagnostics)
            {
                var offsetPair = GetOffset(diagnostic);
                var appliedSuppressions = new List<RuleSuppression>();
                foreach (RuleSuppression suppression in ruleSuppressions)
                {
                    // Check that the diagnostic is within the suppressed extent
                    if (offsetPair != null && (offsetPair.Item1 < suppression.StartOffset || offsetPair.Item2 > suppression.EndOffset))
                    {
                        continue;
                    }

                    // If there's a rule suppression ID, check that it matches the diagnostic
                    if (!string.IsNullOrWhiteSpace(suppression.RuleSuppressionID)
                        && !string.Equals(diagnostic.RuleSuppressionID, suppression.RuleSuppressionID, StringComparison.OrdinalIgnoreCase))
                    {
                        continue;
                    }

                    suppression.Kind = RuleSuppression.SuppressionKind.InSource;
                    unappliedSuppressions.Remove(suppression);
                    appliedSuppressions.Add(suppression);

                }

                if (appliedSuppressions.Count > 0)
                {
                    suppressedRecords.Add(new SuppressedRecord(diagnostic, appliedSuppressions));
                }
                else
                {
                    unsuppressedRecords.Add(diagnostic);
                }

            }
            // Do any error reporting for misused RuleSuppressionIDs here
            foreach (RuleSuppression unappliedSuppression in unappliedSuppressions)
            {
                string suppressionIDError = String.Format(Strings.RuleSuppressionIDError, unappliedSuppression.RuleSuppressionID);
                // checks whether are given a string or a file path
                if (String.IsNullOrWhiteSpace(diagnostics.First().Extent.File))
                {
                    unappliedSuppression.Error = String.Format(
                        CultureInfo.CurrentCulture,
                        Strings.RuleSuppressionErrorFormatScriptDefinition,
                        unappliedSuppression.StartAttributeLine,
                        suppressionIDError);
                }
                else
                {
                    unappliedSuppression.Error = String.Format(
                        CultureInfo.CurrentCulture, 
                        Strings.RuleSuppressionErrorFormat, 
                        unappliedSuppression.StartAttributeLine,
                        System.IO.Path.GetFileName(diagnostics.First().Extent.File),
                        suppressionIDError);
                }

                errorRecords.Add(
                    new ErrorRecord(
                        new ArgumentException(unappliedSuppression.Error),
                        unappliedSuppression.Error,
                        ErrorCategory.InvalidArgument, 
                        unappliedSuppression));
                
            }